Add ability to switch off the animation of showons

regularlabs opened the pull request:
### Summary of Changes
To conditionally hide fields or content based on the values of other fields, you can add a `showon=""` attribute to fields in xml files.

However, this will always use a slideUp/slideDown animation.
There are cases where you do not want to animate the showing/hiding of your content or fields.

This patch adds the ability to switch off the animation by adding a class`no-animation` or `no-animate` to the field.
You can also add the class to the surrounding showon container, which can be helpful when creating your own via php.

### Testing Instructions
For instance, you can test with the Joomla core configuration:
```
		<field
			name="page_heading"
			type="text"
			label="COM_MENUS_ITEM_FIELD_PAGE_HEADING_LABEL"
			description="COM_MENUS_ITEM_FIELD_PAGE_HEADING_DESC"
			default=""
			showon="show_page_heading:1"
		/>
```
And change that to:
```
		<field
			name="page_heading"
			type="text"
			label="COM_MENUS_ITEM_FIELD_PAGE_HEADING_LABEL"
			description="COM_MENUS_ITEM_FIELD_PAGE_HEADING_DESC"
			default=""
			showon="show_page_heading:1"
			class="no-animation"
		/>
```
Now go to: `administrator/index.php?option=com_config&view=component&component=com_menus` and toggle the `Show Page Heading` setting.
Before the patch it should have the slide animation. After, it should not animate.

### Documentation Changes Required
https://docs.joomla.org/Form_field
Optionally add a note somewhere that you can switch off the animation effect via the classname.

regularlabs committed:
Adds ability to control the animation of the showon

regularlabs committed:
Adds support for showon-animate="false" in renderField

regularlabs committed:
Adds animate value to parseShowOnConditions output

regularlabs committed:
Minisfies new javascript code in cms.js

regularlabs committed:
Update default.php

regularlabs committed:
Update filters.php

regularlabs committed:
Adds missing comment string

regularlabs committed:
Update cms-uncompressed.js

regularlabs committed:
Update cms.js

regularlabs committed:
Fixes code styling

regularlabs committed:
Fixes code styling

Quy commented on media/jui/js/cms-uncompressed.js:
```diff
@@ -137,6 +141,10 @@ Joomla = window.Joomla || {};
 				}
 			}
 
+			if(typeof animate === "undefined") {
```
```suggestion
			if (typeof animate === "undefined") {
```

Quy commented on media/jui/js/cms-uncompressed.js:
```diff
@@ -61,6 +61,10 @@ Joomla = window.Joomla || {};
 				fieldName = condition.field;
 				$fields   = $('[name="' + fieldName + '"], [name="' + fieldName + '[]"]');
 
+				if(typeof animate === "undefined" && typeof condition.animate !== "undefined") {
```
```suggestion
				if (typeof animate === "undefined" && typeof condition.animate !== "undefined") {
```

Commit:
Update media/jui/js/cms-uncompressed.js

Co-Authored-By: regularlabs <info@regularlabs.com>

Commit:
Update media/jui/js/cms-uncompressed.js

Co-Authored-By: regularlabs <info@regularlabs.com>

brianteeman commented:
It does what it says but I dont see the use case??

regularlabs commented:
The slide animation is vertical.
There are cases when this messes up the layout during animation.
Especially when using custom fields that have different layouts than the standard label=>field.

regularlabs commented:
With animation:
![with-animation](https://user-images.githubusercontent.com/13553242/47758147-fe5ea780-dca9-11e8-866d-a93cb9c59876.gif)

Without Animation:
![no-animation](https://user-images.githubusercontent.com/13553242/47758145-facb2080-dca9-11e8-8848-3bbb4b4a4514.gif)

brianteeman commented:
Thanks

brianteeman commented:
I have tested this item :white_check_mark: successfully on 8f31b56fa9ed6b2ffafe68a62d7637b1911626a5<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22870">issues.joomla.org/tracker/joomla-cms/22870</a>.</sub>

regularlabs commented:
Also there are cases when you have a LOT of fields hiding/showing based on a single value.
And then animation can cause the page to react very slowly (especially on older browsers).

Also, visually animation it is not always what you want.
For instance, when you want to toggle 2 fields (same label & options, just different field name) without the user seeing it is actually a different field.
For instance, an active and a readonly (disabled) version of the field.

bembelimen commented:
Any chance to offer more possibilities as value for this parameter? "none", "vertical", "horizontal", ...

dgrammatiko commented:
FWIW in Joomla 4 there ARE NO animations, so introducing this feature in tagging means that the project should keep the same API (with the redundant animation part) also in J4. @wilsonge keep a note on this please...

regularlabs commented:
@bembelimen 
There is no ready-to-use horizontal slide effects in the jQuery version included in Joomla 3.
So I think it is beyond the scope of this PR to implement that.

regularlabs commented:
Any reason why this isn't merged yet?

dgrammatiko commented:
@regularlabs I guess because of 
https://github.com/joomla/joomla-cms/pull/22870#issuecomment-434618117

regularlabs commented:
No, because of:
https://github.com/joomla/joomla-cms/pull/22912#issuecomment-438895546

dgrammatiko commented:
Honestly it will be ultra stupid to introduce an API change in 3.9 and remove it in 4.0...

regularlabs commented:
It is solving an issue we have in Joomla 3.
If Joomla 4 solves that in another fashion (by having no animations at all), then fine.
Joomla 4 is changing behaviour anyway.

dgrammatiko commented:
@regularlabs I think you miss my point, you can implement this without adding technical debt for J4 by using a class, eg `no-animation`. This will have minimal impact for J4, in contrary the introduction of another attribute if that needs to be removed in J4 brings a B/C break...

regularlabs committed:
Merge remote-tracking branch 'remotes/upstream/staging' into patch-1

regularlabs committed:
Removes the showon-animate attributes and bases animation on whether element has class no-animate

regularlabs committed:
Removes the showon-animate attributes and bases animation on whether element has class no-animate

regularlabs committed:
Merge remote-tracking branch 'origin/patch-1' into patch-1

regularlabs committed:
Reverts syntax changes in other files

regularlabs committed:
Adds checks for no-animation class on form elements too (not just showon container div)

regularlabs commented:
I totally agree with you!

I have changed the code, which means that only the cms.js file has now changed.
No php changes. Not API changes in the xml elements.

I also updated the description of the PR.

So I guess the RTC tag needs to get removed and we need to do testing again.

If it is better to trash this PR and create a new fresh one, let me know.

zero-24 commented:
tanking RTC off as requested.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22870">issues.joomla.org/tracker/joomla-cms/22870</a>.</sub>

regularlabs committed:
Merge upstream

regularlabs committed:
Merge upstream

regularlabs commented:
@Quy Could you test again? See if it is still working as it should?

Quy commented:
I have tested this item :white_check_mark: successfully on a85f10c45fc84b338c69dc0ad86b928e0ad93f4b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22870">issues.joomla.org/tracker/joomla-cms/22870</a>.</sub>

dgrammatiko commented:
I have tested this item :white_check_mark: successfully on a85f10c45fc84b338c69dc0ad86b928e0ad93f4b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22870">issues.joomla.org/tracker/joomla-cms/22870</a>.</sub>

Quy commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22870">issues.joomla.org/tracker/joomla-cms/22870</a>.</sub>

regularlabs commented:
👍

C-Lodder commented:
1 thing I would have suggested in this is not to use `toggle()` as it was deprecated in 1.8 and remove in 1.9. The only reason it works is because you ship jquery-migrate, however there are some optimisation plugins that allow you to unset it from the document and thus `toggle()` will be undefined.

Should have used `.css(...)` instead.

regularlabs commented:
The existing js code there was already using toggle(). This PR hasn't changed that.
